### PR TITLE
Update BadgeIndicatorContainer visibility when BadgeView is created

### LINKED ISSUE #638: [Bug] BadgeView BadgeIndicatorContainer is visible if the Text property is 0 at the start
### Description
When showing a page containing a  `BadgeView ` with it Text property is initially equals to 0, the badge is visible with "0" as text while it shouldn't be.

### Steps to Reproduce

1. In the sample project change [`Counter` initial value](https://github.com/xamarin/XamarinCommunityToolkit/blob/1860937eacde3ced52e32fffff08a7592e8754b1/src/CommunityToolkit/Xamarin.CommunityToolkit.Sample/ViewModels/Views/BadgeViewViewModel.cs#L10) to 0.

### Expected Behavior
The BadgeIndicatorContainer should not be visible.
### Actual Behavior
The BadgeIndicatorContainer  is visible with value 0.
### Basic Information

- Version with issue:
- Last known good version: NA
- IDE: VS2019
- Platform Target Frameworks: ALL
- Android Support Library Version: NA
- Nuget Packages: NA
- Affected Devices: NA

### LINKED ISSUE #640: [Bug] BadgeView doesn't display when Text is empty
### Description
BadgeView doesn't display when initial Text value is null or empty.

### Steps to Reproduce

1. Have a badgeview with Content.
2. Set its Text to string.Empty or null
3. Watch the badgeview and its content be IsVisible = false

### Expected Behavior
The content is visible, the badge is not as there is no value.

### Actual Behavior
There is no content

### Basic Information

- Version with issue: Commit 4f6f80e9
- Last known good version: never
- IDE: VS 19
- Platform Target Frameworks: 
  - iOS:  
  - Android:  
  - UWP:   
- Android Support Library Version: 
- Nuget Packages:
- Affected Devices:

### Reproduction imagery

Not overly helpful as the visual is to be invisible.

### Reproduction Link

Add to BadgeSamplePage
```
                <xct:BadgeView
                    BackgroundColor="Red"
                    Text=""
                    TextColor="White">
                    <Label Text="Initial Empty Value" />
                </xct:BadgeView>
```

### PR DESCRIPTION
**Pull request by Cfun1**
### Description of Change ###
Initially set `IsInvisible="true"` for `BadgeIndicatorContainer` so that it will be invisible when `BadgeView` is created and it `Text` property is initially equals to 0.

### Bugs Fixed ###
- Fixes #640

Screenshot_1607114251

### Behavioral Changes ###
The `BadgeIndicatorContainer` will be invisible if when `BadgeView` is created it `Text` property is initially equals to 0.

-  For some reasons the badge position is changed after the modifications in this PR from top right to top left.

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [ ] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard
- [ ] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

### CODE REVIEW

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 1]
```diff
@@ -184,7 +184,8 @@ static Grid CreateIndicatorContainerElement()
 		   => new Grid
 		   {
 			   HorizontalOptions = LayoutOptions.Start,
-			   VerticalOptions = LayoutOptions.Start
+			   VerticalOptions = LayoutOptions.Start,
+   			   IsVisible = false
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:188]:
```suggestion
			   IsVisible = false
```

**Review by AndreiMisiukevich — COMMENTED** [review]
LGTM

**Comment by Cfun1** [comment]
@AndreiMisiukevich bindable property name  has the same name as the enum Type should it be renamed? https://github.com/xamarin/XamarinCommunityToolkit/blob/1860937eacde3ced52e32fffff08a7592e8754b1/src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs#L29
 but this is not what causing the position to change.

**Comment by AndreiMisiukevich** [comment]
No, let's leave as is

I think we need to call ```UpdateBadgeViewPlacement()``` in OnTextChanged before UpdateLayout
It should solve our problem

**Comment by Cfun1** [comment]
Exactly what I was about to test right now :)

**Comment by AndreiMisiukevich** [comment]
Oh, probably it won't help(

Seem we should observe ```BadgeIndicatorContainer.IsVisible``` and when this property changes we will have to call ```UpdateBadgeViewPlacement(true)``

**Comment by Cfun1** [comment]
I tested you are right It becomes visible before changing the position

**Comment by Cfun1** [comment]
> 
> 
> Oh, probably it won't help(
> 
> Seem we should observe `BadgeIndicatorContainer.IsVisible` and when this property changes we will have to call `UpdateBadgeViewPlacement(true)`

I tried with (also with `PropertyChanged `)
```
		static Grid CreateIndicatorContainerElement()
		{
			var newgrid = new Grid
			{
				HorizontalOptions = LayoutOptions.Start,
				VerticalOptions = LayoutOptions.Start,
				IsVisible = false
			};
			newgrid.PropertyChanging += IndicatorContainerElement_PropertyChanged;
			return newgrid;
		}

		public static void IndicatorContainerElement_PropertyChanged(object sender, PropertyChangingEventArgs e)
		{
			if (e.PropertyName.Equals(nameof(IsVisible)))
				((sender as Grid).Parent as BadgeView)?.UpdateBadgeViewPlacement(true);
		}
```

Unfortunately it is not working maybe  because `UpdateVisibilityAsync` is running async or I am missing something else.

**Comment by Cfun1** [comment]
I am for now testing without animation only you may need to set  `IsAnimated="False"` in xaml.

Another suggestion is to call `UpdateBadgeViewPlacement` like:

From https://github.com/xamarin/XamarinCommunityToolkit/blob/1860937eacde3ced52e32fffff08a7592e8754b1/src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs#L351-L352

To:
```
else
{
		BadgeIndicatorContainer.IsVisible = badgeIsVisible;
		await UpdateBadgeViewPlacement(true);
 }
```

it is updating the position but for some reasons the dimensions are not good (small) when incrementing first time from 0 up to 3, after 4 it becomes good.

**Comment by Cfun1** [comment]
It seems working as expected, now is ready to be tested.

**Comment by Cfun1** [comment]
Ignore the last commit as it appears it creating an infinite loop when badge position is other than TopRight.

**Comment by Cfun1** [comment]
I have changed below line for debugging

https://github.com/xamarin/XamarinCommunityToolkit/blob/1860937eacde3ced52e32fffff08a7592e8754b1/src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs#L255
To
`if ((BadgeContent.Height <= 0 && BadgeContent.Width <= 0) || (BadgeText.Height <= 0 && BadgeText.Width <= 0))`

and found that `BadgeText.Height` and `BadgeText.Width` equals -1 while it is visible and at the same time `BadgeContent.Width` has a correct value. I don't know why `BadgeText` are not updated on the on the first increment of text value from 0 to 1.

With the modification mentioned in this comment and by ignoring the last commit, the badge position is updated but only after 2 textchanged events (means from value 0 to 2)

**Comment by Cfun1** [comment]
I have changed below line for debugging

https://github.com/xamarin/XamarinCommunityToolkit/blob/1860937eacde3ced52e32fffff08a7592e8754b1/src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs#L255
To
`if ((BadgeContent.Height <= 0 && BadgeContent.Width <= 0) || (BadgeText.Height <= 0 && BadgeText.Width <= 0))`

and found that `BadgeText.Height` and `BadgeText.Width` equals -1 while it is visible and at the same time `BadgeContent.Width` has a correct value. I don't know why `BadgeText` are not updated on the on the first increment of text value from 0 to 1.

Status now: The badge position is updated but only after 2 TextChanged events (means from value 0 to 2)

**Comment by AndreiMisiukevich** [comment]
Hmm, interesting case. I'll take a look and try to fix myself (I am not sure I will have success haha)

**Comment by AndreiMisiukevich** [comment]
I guess this badge should stick to the top right corner
image

Now it's a little above

**Comment by AndreiMisiukevich** [comment]
@Cfun1 https://github.com/xamarin/XamarinCommunityToolkit/pull/654

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 2]
```diff
@@ -230,7 +231,7 @@ void UpdateLayout()
 			BadgeIndicatorBackground.BorderColor = BorderColor;
 			BadgeIndicatorBackground.HasShadow = HasShadow;
 
-			BadgeText.Text = Text;
+			BadgeText.Text = string.IsNullOrEmpty(Text) ? "0" : Text;
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:234]: I don't think it's the proper way.
I think we should just delete the check inside UpdateVisibilityAsync

Comment (Cfun1) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:234]: I have tried with https://github.com/xamarin/XamarinCommunityToolkit/blob/1860937eacde3ced52e32fffff08a7592e8754b1/src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs#L325-329 commented out but it was not working.

**Comment by Cfun1** [comment]
> 
> 
> I guess this badge should stick to the top right corner
> image
> 
> Now it's a little above

true, I am thinking we need to change the way how margins are calculated.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 3]
```diff
@@ -268,7 +269,7 @@ void UpdateBadgeViewPlacement(bool force = false)
 			{
 				const double Padding = 6;
 				var size = Math.Max(BadgeText.Height, BadgeText.Width) + Padding;
-				BadgeIndicatorBackground.HeightRequest = size;
+				BadgeIndicatorBackground.HeightRequest = BadgeText.Height + Padding;
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:272]: do we need to add Padding twice?

**Comment by AndreiMisiukevich** [comment]
#638 is fixed

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 4]
```diff
@@ -272,7 +272,7 @@ void UpdateBadgeViewPlacement(bool force = false)
...
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:275]: it doesn't seem to be a proper fix cuz indicator should be stuck to the corner. Now it's above

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs (hunk 2 above)
Comment (AndreiMisiukevich) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:237]: maybe let's set here Text as it was before, but maybe we will hide indicator if the Value is "0" or null or emtpy?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs (hunk 4 above)
Comment (AndreiMisiukevich) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:275]: Also, currently, it's supposed to have equal width and height. So for allowing non-square / non-circle shapes, I think, we should expose a new property

**Comment by AndreiMisiukevich** [comment]
Please, focus on #640 in this PR :)
Does it sound OK for you?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs (hunk 4 above)
Comment (Cfun1) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:275]: I think current margins calculations are supposing to not have a big difference between Width and Height.

Comment (AndreiMisiukevich) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:275]: ``var size = Math.Max(BadgeText.Height, BadgeText.Width) + Padding```
It always takes "greatest" parameter now

**Comment by Cfun1** [comment]
Of course no issue :)

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 5]
```diff
@@ -328,12 +328,13 @@ async Task UpdateVisibilityAsync()
...
 			{
-				IsVisible = false;
+				BadgeIndicatorBackground.IsVisible = false;
 				return;
 			}
 
 			var badgeIsVisible = !AutoHide || !badgeText.Trim().Equals("0");
-
+			BadgeIndicatorBackground.IsVisible = badgeIsVisible;
+
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:337]: Seems this line is redundant (We set badgeIsVisible below)

Comment (Cfun1) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:337]: we set for `BadgeIndicatorContainer` not for `BadgeIndicatorBackground` in line 357. If this line removed together with 331 then a red frame will be visible.

**Review by jfversluis — APPROVED** [review]

**Pull request merged**